### PR TITLE
Support for using a subpath

### PR DESCRIPTION
**Pull request by matthewp**
## Changes

- When using `astro dev` if you have a `buildOptions.site` set with a subpath, that will be used as the root.
- If you attempt to go to `/` you will see this 404 page:
    Screen Shot 2021-11-11 at 4 57 22 PM
- The console gives you this:
    Screen Shot 2021-11-11 at 4 57 53 PM
- This is going to be updated when a better 404 page is designed: https://discord.com/channels/830184174198718474/854834010642579457/908465177584894012
- When you use `astro preview` you get the same behavior.
- The build still builds to `dist/` so you can deploy that folder to your host file system.

## Testing

- This touches the dev server and astro preview. An example app was created to verify those are behaving correctly.
- There are existing unit tests that verify that relative paths are created by the build.

### CODE REVIEW

**Review by jonathantneal — APPROVED** [review]
I like it. The message page with the redirect link seems convenient, too.

**Comment by natemoo-re** [comment]
Looks great!

**Comment by matthewp** [comment]
Note that I need to verify that resources in `public/` resolve to the correct place.

**Comment by drwpow** [comment]
Shoot—merged the 404 page restyling not knowing this PR also used the dev template. I separated the `4xx` and `5xx` errors but feel free to rearrange anything as you see fit. My feeling is that a few templates may be easier to manage than one to rule them all

File: packages/astro/src/core/dev/index.ts [hunk 1]
```diff
@@ -347,7 +362,11 @@ export class AstroDevServer {
     }
     // if not found, fall back to default template
     else {
-      html = errorTemplate({ statusCode, title: 'Not found', tabTitle: '404: Not Found', message: pathname });
+      if(pathname === '/' && !pathname.startsWith(this.pathname)) {
+        html = subpathNotUsedTemplate(this.pathname);
+      } else {
+        html = errorTemplate({ statusCode, title: 'Not found', tabTitle: '404: Not Found', message: pathname });
```
Comment (drwpow) [inline packages/astro/src/core/dev/index.ts:368]: Slight suggestion: you could probably add a new conditional param to the 4xx template to conditionally show this message or not (also happy to pair on this if needed)

**Pull request merged**